common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,287 @@
+    {
+      if (!g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return FALSE;
+        }
+    }
+
+  generator = json_generator_new ();
+  json_generator_set_root (generator, node);
+  json_generator_set_pretty (generator, FALSE);
+
+  /* Resize the buffer to accomadate one extra newline and write that
+   * in place at serialized_len */
+  serialized = json_generator_to_data (generator, &serialized_len);
+  serialized = g_realloc (serialized, (serialized_len + 1) * sizeof (gchar));
+  serialized[serialized_len++] = '\n';
+  serialized[serialized_len] = '\0';
+
+  stream = g_file_append_to (file,
+                             G_FILE_CREATE_NONE,
+                             cancellable,
+                             error);
+
+  return g_output_stream_write_all (G_OUTPUT_STREAM (stream),
+                                    (gconstpointer) serialized,
+                                    serialized_len,
+                                    &written,
+                                    cancellable,
+                                    error);
+}
+
+#define EVENT_LOG_VERSION "1"
+
+static JsonNode *
+build_base_event (const gchar *event)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+
+  json_builder_begin_object (builder);
+
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_set_member_name (builder, "version");
+  json_builder_add_string_value (builder, EVENT_LOG_VERSION);
+  json_builder_end_object (builder);
+
+  return json_builder_get_root (builder);
+}
+
+static JsonNode *
+deploy_install_event (const gchar *ref,
+                      const gchar *origin)
+{
+  JsonNode *base = build_base_event ("deploy-install");
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+  json_object_set_string_member (object, "origin", origin);
+
+  return base;
+}
+
+/**
+ * flatpak_transaction_log_commit_deploy_install_event:
+ * @log: a #FlatpakTransactionLog
+ * @ref: a decomposable Flatpak ref
+ * @origin: the origin of the flatpak to be installed, whether it is from
+ *          a bundle or from a remote.
+ * @cancellable: a #GCancellable
+ * @error: a #GError
+ *
+ * Write a "deploy-install" event to the transaction log. The "ref" and
+ * "origin" arguments will be included in the event description as properties.
+ *
+ * Returns: %TRUE if writing the event succeeded.
+ */
+gboolean
+flatpak_transaction_log_commit_deploy_install_event (FlatpakTransactionLog *log,
+		                                                 const gchar           *ref,
+		                                                 const gchar           *origin,
+		                                                 GCancellable          *cancellable,
+		                                                 GError                **error)
```

Comment:
Nitpick: looks like the argument alignment is off (unless it’s being messed up by github?)

Reply:
If you're referring to the alignment of `**error` - is the convention here that it is shifted back by one (such that the starting character of the identifier aligns?)

Reply:
Ah, looks like the tabs monster strikes again :(

Reply:
Yes, the starting characters of all the identifiers should align. And there should be no tabs; just spaces.

Reply:
👾